pennylane/devices/default_clifford.py

```diff
@@ -519,58 +692,322 @@ def _measure_state(self, tableau_simulator, circuit, global_phase):
+            )
+
+        if self._prob_states is None:
+            num_wires = len(meas_wires)
+            basis_vec = np.arange(2**num_wires)[:, np.newaxis]
+            self._prob_states = (((basis_vec & (1 << np.arange(num_wires)[::-1]))) > 0).astype(int)
+
+        # TODO: We might be able to skip the inverse done below
+        # (as the distribution should be independent of inverse)
+        diagonalizing_cit = kwargs.get("stim_circuit").copy()
+        diagonalizing_ops = [] if not meas.obs else meas.obs.diagonalizing_gates()
+        for diag_op in diagonalizing_ops:
+            # Check if it is Clifford
+            if diag_op.name not in _GATE_OPERATIONS:  # pragma: no cover
+                raise ValueError(
+                    f"Currently, we only support observables whose diagonalizing gates are Clifford, got {diag_op}"
+                )
+            # Add to the circuit to rotate the basis
+            stim_op = _pl_to_stim(diag_op)
+            if stim_op[0] is not None:
+                diagonalizing_cit.append_from_stim_program_text(f"{stim_op[0]} {stim_op[1]}")
+
+        # Build the Tableau simulator from the diagonalized circuit
+        circuit_simulator = stim.TableauSimulator()
+        circuit_simulator.do_circuit(diagonalizing_cit)
+        if not self._tableau:
+            state = self._measure_state(meas, circuit_simulator, circuit=circuit)
+            return meas.process_state(state, wire_order=circuit.wires)
+
+        tgt_states = self._prob_states
+        if not tgt_states.shape[1]:
+            raise ValueError("Cannot set an empty list of target states.")
+
+        if len(meas_wires) < tgt_states.shape[1]:
+            tgt_states = []
+            for state in self._prob_states:
+                if list(state[meas_wires]) not in tgt_states:
+                    tgt_states.append(list(state[meas_wires]))
+            tgt_states = np.array(tgt_states)
+        else:
+            meas_wires = meas_wires[: tgt_states.shape[1]]
+
+        # Maintain a representaiton of basis states to build a visit-array
+        tgt_integs = np.array([int("".join(map(str, tgt_state)), 2) for tgt_state in tgt_states])
+
+        # Iterate over the required basis states and for each of them compute the probability
+        # If an impossible branch occur keep a note of it via a visit-array
+        # Worst case complexity O(B * M * N), where N is #measured_qubits and B is basis states.
+        visited_probs = []
+        prob_res = np.ones(tgt_states.shape[0])
+        for tgt_index, (tgt_integ, tgt_state) in enumerate(zip(tgt_integs, tgt_states)):
+            if tgt_integ in visited_probs:
+                continue
+            prob_sim = circuit_simulator.copy()
+            for idx, wire in enumerate(meas_wires):
+                expectation = prob_sim.peek_z(wire)
+                # (Eig --> Res) | -1 --> 1 | 1 --> 0 | 0 --> 0 / 1 |
+                outcome = int(0.5 * (1 - expectation))
```

Comment:
is the `0 --> 0 / 1` randomness from rounding `0.5`, and if so what's the consequence of that?